sdk/sdk/src/main/java/org/apache/carbondata/sdk/file/PaginationCarbonReader.java

```diff
@@ -0,0 +1,276 @@
+    return getRows(from, to);
+  }
+
+  /**
+   * Get total rows in the folder.
+   * It is based on the snapshot of files taken while building the reader.
+   *
+   * @return total rows from all the files in the reader.
+   */
+  public long getTotalRows() {
+    if (isClosed) {
+      throw new RuntimeException("Pagination Reader is closed. please build again");
+    }
+    return rowCountInSplits.get(rowCountInSplits.size() - 1);
+  }
+
+  /**
+   * This interface is for python to call java.
+   * Because python cannot understand java Long object. so send string object.
+   *
+   * Get total rows in the folder.
+   * It is based on the snapshot of files taken while building the reader.
+   *
+   *
+   * @return total rows from all the files in the reader.
+   */
+  public String getTotalRowsAsString() {
+    if (isClosed) {
+      throw new RuntimeException("Pagination Reader is closed. please build again");
+    }
+    return (rowCountInSplits.get(rowCountInSplits.size() - 1)).toString();
+  }
+
+  private static int findBlockletIndex(List<Long> summationArray2, Long key) {
+    // summation array is in sorted order, so can use the binary search.
+    int index = Collections.binarySearch(summationArray2, key);
+    if (index < 0) {
+      // when key not found, binary search returns negative index [-1 to -N].
+      // which is the possible place where key can be inserted.
+      // with one shifted position. As 0 is also a valid index.
+      index = index + 1; // offset the one index shifted.
+      index = Math.abs(index);
+    }
+    return index;
+  }
+
+  private int[] getBlockletIndexRange(long from, long to) {
+    // find the matching blocklet index range by checking with from and to.
+    int upperBound = findBlockletIndex(rowCountInSplits, to);
+    // lower bound cannot be more than upper bound, so work on sub list.
+    int lowerBound = findBlockletIndex(rowCountInSplits.subList(0, upperBound), from);
+    return new int[] {lowerBound, upperBound};
+  }
+
+  private Object[] getRows(long from, long to)
+      throws IOException, InterruptedException {
+    int rowCount = 0;
+    Object[] rows = new Object[(int)(to - from + 1)];
+    // get the matching split index (blocklets) range for the input range.
+    int[] blockletIndexRange = getBlockletIndexRange(from, to);
+    for (int i = blockletIndexRange[0]; i <= blockletIndexRange[1]; i++) {
+      String blockletUniqueId = String.valueOf(i);
+      BlockletRows blockletRows;
+      if (cache.get(blockletUniqueId) != null) {
+        blockletRows = (BlockletRows)cache.get(blockletUniqueId);
+      } else {
+        BlockletDetailInfo detailInfo =
+            ((CarbonInputSplit) allBlockletSplits.get(i)).getDetailInfo();
+        int rowCountInBlocklet = detailInfo.getRowCount();
+        Object[] rowsInBlocklet = new Object[rowCountInBlocklet];
+        // read the rows from the blocklet
+        // TODO: read blocklets in multi-thread if there is a performance requirement.
```

Comment:
How about the performance for pagination?

Reply:
Tested with local file system. With cache reading few rows (out of million rows) after cache is loaded is very fast (within 1 second)